src/Build/Evaluation/Evaluator.cs

```diff
@@ -1428,7 +1428,7 @@ private void EvaluateImportElement(string directoryOfImportingFile, ProjectImpor
         {
             using (_evaluationProfiler.TrackElement(importElement))
             {
-                List<ProjectRootElement> importedProjectRootElements = ExpandAndLoadImports(directoryOfImportingFile, importElement, out var sdkResult);
+                IEnumerable<ProjectRootElement> importedProjectRootElements = ExpandAndLoadImports(directoryOfImportingFile, importElement, out var sdkResult);
 
                 foreach (ProjectRootElement importedProjectRootElement in importedProjectRootElements)
```

Comment:
Changing the type of the variable to `IEnumerable` has the unfortunate implication of regressing `foreach` performance. Compare the IL code here:

https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKAwAYACDHAFgG5CMBmcrUgYVIG9DTvzGNLSAWQAUAGSgBnBAB4KxAHykUkhAEoOXHloBmAewBOAUwgBjABalhc0hNKwlK1Zq3dOBFx/I4AnMImrad08AX2dSUKDuBnIBADkrHHpZHAUHKXU3T1I9I1MLBLJbe2V0sK1MrJ4KX39ArIitCOCgA

When typed as `IEnumerable`, List's enumerator is boxed (an extra allocation) and then accessed through the `IEnumerator` interface (tiny bit slower than direct calls).

Can this optimization be implemented without changing the type, by using a special value of `null` to represent an empty list for example?

Reply:
I know about value type based enumerator for `List<T>` and I was thinking about using `null` to keep specific type. To be honest, I don't know why I rejected this idea so it worth to use it. 
I will try to use `ImmutableArray` and `MoveToImmutable` as Drew mentioned. Maybe we will not need `List` at all.

Reply:
@ladipro I changed the code to use `null`s.